### ISSUE TITLE
Multipath config for TrueNAS

### ISSUE THREAD

**Issue by jaronparsons** [xcp-ng/xcp#297]
Using the default /etc/multipath.conf included with xcp-ng results in the multipathing flipping between the active and passive controllers on a TrueNAS system. The internal network between the controllers is a 1Gbps link.  This can lead to saturation issues when traffic is communicating through the passive controller to the active. Especially when connected via 10Gbps to the storage network.
Add the following to /etc/multipath.conf and restart the multipathd service to resolve this issue.

```
device {
                vendor                  "TrueNAS"
                product                 "iSCSI Disk"
                hardware_handler        "1 alua"
                path_grouping_policy    group_by_prio
                prio                    alua
        }
```

It would be nice if this could be added as a default to /etc/multipath.conf

**Comment by olivierlambert** [comment]
Hi @jaronparsons 

I suppose you are working at *iXsystems Inc.* right?

**Comment by jaronparsons** [comment]
correct
IT ;)

**Comment by olivierlambert** [comment]
Okay. I'm assigning @stormi to make the modification so we can embed this in a next update (and also next XCP-ng version obviously). I assume we can trust you on that. Ideally, we'd like to create an test update so you can validate it before it's going in our public repo. What do you think?

Also, we'd like to have a direct communication with you guys, for such requests but also to work closely together (there's a lot of potential synergies).

**Comment by jaronparsons** [comment]
Sounds great!
I have currently deployed this fix on 20+ xcp-ng 8.0.0 servers and have been running for the last 2+ months to ensure stability and compatibility.  Is that enough of a test?  
I am not much not much of a Dev, but I am IT Manager and report directly to Engineering.  I can get you contact with the correct people at anytime ;)  Shoot me an E-mail and we can open a dialog if you like and I can loop in our Engineering and support people if you like.

**Comment by olivierlambert** [comment]
That's enough indeed :stuck_out_tongue: 

Send me an email to contact at xcp dash ng dot com, we'd like to have a channel opened between us to work together in the right direction. Thanks :+1:

**Comment by stormi** [comment]
The main work will be to find where to do the change.

The `/etc/multipath.conf` file does not belong to any RPM of the system (not a very good practice), so probably created either as a `%post` RPM scriptlet, or during system installation, or written by some script called by XAPI. I'll need to dig.

There's also a similar `/etc/multipath.xenserver/multipath.conf` file that belongs to the `sm` RPM. I don't know yet what it's for.

**Comment by jaronparsons** [comment]
I just noticed that /etc/multipath.conf is actually linked to  /etc/alternatives/multipath.confAlso that the changes I put in the file automatically appear in the /etc/multipath.xenserver/multipath.conf

**Comment by nagilum99** [comment]
@stormi To get our HPE MSA2040 SAN FC working, I added the corresponding entries to /etc/multipath.conf which should belong to multipathd (I assume the package has the same name?)

I never touched any other files, also it would be a good idea to get it upstream - if they accept it.
If not: Do your own thing. ;>

**Comment by stormi** [comment]
@nagilum99 which corresponding entries? The same as that from @jaronparsons?

**Comment by stormi** [comment]
@jaronparsons Could you try to contribute your changes as a pull request to https://github.com/xapi-project/sm/blob/master/multipath/multipath.conf? This is Citrix's upstream git repository for the `sm` package which contains the `multipath.conf` file.

I'm also working on adding the change to our own `sm` package, but we always try to push changes upstream.

**Comment by stormi** [comment]
So, I can easily add the change to our `sm` package, **but** I'm not confident to do it before the next release of XCP-ng (8.1), because the `sm` package is made in a way that changing `multipath.conf` will cause any user-made changes to the file to be lost. I can't risk breaking existing setups with an update.

For those who can `speak` RPM spec file language, it means that `/etc/multipath.xenserver/multipath.conf` is defined with `%config` rather than `%config(noreplace)`. I could change that too, but then I'd have to revert that specific change in XCP-ng 8.1 so that I don't drift too far away from the Citrix package in the long term.

So my proposal is to wait for XCP-ng 8.1. The classical upgrade method overwrites that kind of configuration changes, but there's a proper warning in the documentation, documentation that users are more likely to read for an upgrade to a new version than when merely applying updates.

And for those who'll chose the yum upgrade path, we'll need to add a warning in the upgrade docs too, saying that any change brought to multipath.conf will have to be made again.

**Comment by nagilum99** [comment]
> @nagilum99 which corresponding entries? The same as that from @jaronparsons?

No, corresponding to our MSA. I wanted to say: To fix our problem, I didn't touch anything beside /etc/multipath.conf.

**Comment by stormi** [comment]
> @jaronparsons Could you try to contribute your changes as a pull request to https://github.com/xapi-project/sm/blob/master/multipath/multipath.conf? This is Citrix's upstream git repository for the `sm` package which contains the `multipath.conf` file.
> 
> I'm also working on adding the change to our own `sm` package, but we always try to push changes upstream.

Hi @jaronparsons, I'd still like a pull request to Citrix's https://github.com/xapi-project/sm/blob/master/multipath/multipath.conf

**Comment by stormi** [comment]
> And for those who'll chose the yum upgrade path, we'll need to add a warning in the upgrade docs too, saying that any change brought to multipath.conf will have to be made again.

I have added a more generic section about configuration changes in the context of yum-style upgrades: https://github.com/xcp-ng/xcp/wiki/Upgrade-Howto#3-check-the-configuration-files

**Comment by stormi** [comment]
> > @jaronparsons Could you try to contribute your changes as a pull request to https://github.com/xapi-project/sm/blob/master/multipath/multipath.conf? This is Citrix's upstream git repository for the `sm` package which contains the `multipath.conf` file.
> > I'm also working on adding the change to our own `sm` package, but we always try to push changes upstream.
> 
> Hi @jaronparsons, I'd still like a pull request to Citrix's https://github.com/xapi-project/sm/blob/master/multipath/multipath.conf

Ping @jaronparsons?

I'm working on XCP-ng 8.1, a first beta is close, and I'm reluctant to maintain that patch forever in XCP-ng without even trying to submit it upstream.

**Comment by jaronparsons** [comment]
I have submitted.
If it doesnt look right let me know and ill adjust.
Sorry for the delay
--
Jaron Parsons
IXsystems, Inc. <http://www.ixsystems.com>
Manager of IT
Office: (408) 943-4100 ext.351

**Comment by stormi** [comment]
Thanks! I've added a comment to the PR.

**Comment by Fohdeesha** [comment]
hey @jaronparsons we've got what I believe the first XCP-ng customer using ALUA, with this config specifically (he's a truenas customer). Ever since enabling ALUA and this config, he's had issues with his SRs, snapshots, that kind of thing. At one point when he rebooted a VM, it booted up in an older state. We couldn't find any issue on our side, but ALUA is not our specialty (we don't have many customers using ISCSI at all, they've all moved on to NFS). I believe he should be opening a ticket with ixsystems, hopefully we can all figure it out before getting this config implemented into the xcp-ng distro.

**Comment by nagilum99** [comment]
ALUA is nothing iSCSI specific, it's also used for FC (derivates). It just controlls which of the offered channels/ports to use for data transmission.
E.g. 2 controllers, Active/Active - 2 ports each -> 4 in total.
One Controller is always the master and handles all IO of specific LUN/RAID set, so all traffic that goes to the 2nd controller needs to be handled over to the master. This adds latency and possibly (as with TrueNAS) is bottlenecked by a small bandwith interconnect.
So ALUA negotiates with the controller, which paths are the optimized ones and handles traffic just over them, as long as they are 'up'.

ALUA cannot, never ever, do things like give you an old snapshot back, it's simply not it's layer or anything which is under controll of that function. It's just prioritizing/balancing paths from the host to a target.
The problem must be somewhere at TrueNAS or snapshotting functions of that XCP-ng install.

**Comment by jaronparsons** [comment]
Yes it would be best to open a ticket with ixsystems directly.  Our support
dept created this config and should be familiar with it.  This config is
specifically for TrueNAS with dual controllers (as nearly all are) and
simply eliminates the vm/host from access the storage across the internal
interface between the controllers. Support will be able to assist.
--
Jaron Parsons
IXsystems, Inc. <http://www.ixsystems.com>
Manager of IT
Office: (408) 943-4100 ext.351

**Comment by stormi** [comment]
Reference to the pull request: https://github.com/xapi-project/sm/pull/481

**Comment by stormi** [comment]
Closing since the patch was included in XCP-ng 8.1, but we'll continue to watch the outcome of the pull request.
